lib/streamlit/commands/navigation.py

```diff
@@ -0,0 +1,178 @@
+    available pages can be updated with each rerun for dynamic navigation. In
+    this mode, the `pages/` folder is ignored.
+
+    By default, `st.navigation` draws the available pages in the side
+    navigation. This behavior can be changed using the `position=` keyword
+    argument.
+
+    Parameters
+    ----------
+    pages: list[st.Page] or dict[str, list[st.Page]]
+        A list of `st.Page` objects or a dictionary where the keys are section
+        headers and the values are lists of `st.Page` objects.
+
+    position: str
+        The position of the navigation menu. Can be "sidebar" or "hidden".
+
+    Example
+    -------
+    >>> import streamlit as st
+    >>> from pages import page1, page2
+    >>>
+    >>>	pg = st.navigation([st.Page(page1), st.Page(page2)])
+    >>>
+    >>>	st.title("My Awesome App")
+    >>> pg.run()
+    """
+    ctx = get_script_run_ctx()
+    if not ctx:
+        return None
+
+    nav_sections = {"": pages} if isinstance(pages, list) else pages
+    page_list = pages_from_nav_sections(nav_sections)
+
+    if len(page_list) == 0:
+        raise StreamlitAPIException(
+            "`st.navigation` must be called with at least one `st.Page`."
+        )
+
+    defaults = [page for page in page_list if page.default]
+
+    # First assume the first page is the default. We will update this if
+    # we detect that a different page is the default.
+    default_page = page_list[0]
+    if len(defaults) > 1:
+        raise StreamlitAPIException(
+            "Multiple Pages specified with `default=True`. At most one Page can be set to default."
+        )
+    if len(defaults) == 0:
+        default_page.default = True
+    else:
+        default_page = defaults[0]
+
+    msg = ForwardMsg()
+    msg.navigation.position = position
+    msg.navigation.sections[:] = nav_sections.keys()
+    for section in nav_sections:
+        for page in nav_sections[section]:
+            p = msg.navigation.app_pages.add()
+            p.page_script_hash = page._script_hash
+            p.page_name = page.title
+            p.icon = page.icon
+            p.is_default = page.default
+            p.section_header = section
+            p.url_pathname = page.title.replace(" ", "_")
+
+    # Inform our page manager about the set of pages we have
+    page_by_hash = nav_sections_to_page_dict(nav_sections)
+    ctx.pages_manager.set_pages(page_by_hash)
+    managed_page = ctx.pages_manager.get_page_script(default_page._script_hash)
+
+    found_page = None
+    if managed_page is None:
+        send_page_not_found(ctx)
+        found_page = default_page
+    else:
+        managed_page_script_hash = managed_page["page_script_hash"]
+        matching_pages = [
+            p for p in page_list if p._script_hash == managed_page_script_hash
+        ]
+        found_page = matching_pages[0]
+        if len(matching_pages) == 0:
+            send_page_not_found(ctx)
+            found_page = default_page
```

Comment:
it looks like `managed_page` at this point always is the `default_page` from [this line above](https://github.com/streamlit/streamlit/pull/8673/files#diff-cde9e91039649c86b0e33690bf374e49cb91304e359cb681c2f4c4414e0d3859R155). So, I am not sure whether these lines are every executed or do I read it wrong?

Reply:
Ah it's a fallback hash. I made it a keyword argument for clarity.